nixos/modules/services/networking/openconnect.nix

```
@@ -0,0 +1,154 @@
+      gateway = mkOption {
+        description = "Gateway server to connect to.";
+        example = "gateway.example.com";
+        type = types.str;
+      };
+
+      protocol = mkOption {
+        description = ''
+          Protocol to use. One of <literal>anyconnect</literal>,
+          <literal>array</literal>, <literal>nc</literal>,
+          <literal>pulse</literal>, <literal>gp</literal>,
+          <literal>f5</literal> and <literal>fortinet</literal>
+        '';
+        example = "anyconnect";
+        type = types.str;
+      };
+
+      user = mkOption {
+        description = "Username to authenticate with.";
+        example = "example-user";
+        type = types.nullOr types.str;
+      };
+
+      # Note: It does not make sense to provide a way to declaratively
+      # set an authentication cookie, because they have to be requested
+      # for every new connection and would only work once.
+      password = mkOption {
+        description = ''
+          Password to authenticate with. WARNING: To prevent this from being
+          stored world-readable in the Nix store, use
+          <code>passwordFile</code> instead. This is passed to
+          <code>openconnect</code> via the <code>--passwd-on-stdin</code> option.
+        '';
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      # Note
+      passwordFile = mkOption {
+        description = ''
+          File containing the password to authenticate with. This
+          is passed to <code>openconnect</code> via the
+          <code>--passwd-on-stdin</code> option.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect-passwd";
+        type = types.path;
+      };
+
+      certificate = mkOption {
+        description = ''
+          Certificate to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect_certificate.pem";
+        type = types.nullOr types.str;
+      };
+
+      privateKey = mkOption {
+        description = ''
+          Private key to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        example = "/var/lib/secrets/openconnect_private_key.pem";
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      extraConfig = mkOption {
+        description = ''
+          Extra config to be appended to the interface config. It should
+          contain long-format options as would be accepted on the command
+          line by <code>openconnect</code> (see https://www.infradead.org/openconnect/manual.html),
+          separated with newlines. Empty lines and lines where the first
+          non-space character is # are ignored.
+        '';
+        default = "";
+        example = ''
+          # Set compression mode
+          compression=stateless
+
+          no-http-keepalive
+          no-dtls
+        '';
+        type = types.str;
```

```suggestion
        type = types.lines;
```
if it really has to be stringy. would be nicer if this was an RFC 42 `settings` attrset instead, but not sure how this would look. the config format doesn't feel very amenable to being represented like that :/

---

Yeah, I don't think that non-KV pair options like `no-dtls` could be easily represented by Nix expressions

---

`no-dtls` sounds like a boolean... any other examples?

---

In my opinion, the problem with the non-key-value pair flags like --no-dtls is that while a boolean KV pair is an accurate representation, it might still be confusing for users reading the docs seeing that both non-KV and KV options are KV pairs in the module config. Intuitively, you'd also expect a config option like deflate = false; to have an effect on the generated config, but if I understand your reasoning correctly, you are proposing to filter for true values and put the corresponding keys into the configfile without the boolean value.

---

You understand me correctly. It is a trade-off. We do something similar for the `mysql` module. `extraConfig` renders configuration useless from a programmers point of view: you can never introspect and use the value. Decent documentation goes a long way in explaining to the user how to use the module.

---

Yes, that sounds reasonable. I'll implement the corresponding conversion.

---

What do you think of my implementation? I think it makes sense to keep `user`, `password`, `certificate` and `privateKey` as separate options because they probably are the most frequently used ones and we can have type-checking for them this way